File: nahida_bot/gateway/schemas.py
"""Pydantic request and response schemas for the WebAPI."""

import re
from typing import Any, Literal

from pydantic import BaseModel, Field, model_validator


# -- Health ---------------------------------------------------------------


class HealthResponse(BaseModel):
    status: str
    app_name: str
    started: bool


# -- Status ---------------------------------------------------------------


class StatusResponse(BaseModel):
    app: dict[str, Any]
    resources: dict[str, Any]
    services: dict[str, Any]
    usage: dict[str, Any]


# -- Bootstrap ------------------------------------------------------------


class BootstrapResponse(BaseModel):
    app_name: str
    version: str
    api_base: str
    webui_base: str
    auth: dict[str, Any]
    features: list[dict[str, str]]
    server_time: str


# -- Auth -----------------------------------------------------------------


class AuthLoginRequest(BaseModel):
    password: str


class AuthLoginResponse(BaseModel):
    authenticated: bool
    mode: str = "password"
    expires_at: str = ""


class AuthSessionResponse(BaseModel):
    authenticated: bool
    auth_required: bool
    mode: str
    expires_at: str = ""


# -- System Actions -------------------------------------------------------


class SystemActionRequest(BaseModel):
    confirm: bool = False
    reason: str = ""


class SystemActionResponse(BaseModel):
    accepted: bool
    action: str
    mode: str
    message: str


# -- Plugins --------------------------------------------------------------


class PluginSummaryResponse(BaseModel):
    id: str
    name: str
    version: str
    description: str = ""
    state: str
    configured_enabled: bool = True
    path: str
    entrypoint: str
    load_phase: str
    nahida_bot_version: str = ""
    sdk_version: str = ""
    error_message: str = ""
    permissions: dict[str, Any] = Field(default_factory=dict)
    capabilities: dict[str, Any] = Field(default_factory=dict)
    depends_on: list[dict[str, str]] = Field(default_factory=list)
    config_keys: list[str] = Field(default_factory=list)
    config_schema: dict[str, Any] = Field(default_factory=dict)
    has_config: bool = False
    has_instance: bool = False
    has_runtime_api: bool = False


class PluginListResponse(BaseModel):
    plugins: list[PluginSummaryResponse]


class PluginActionResponse(BaseModel):
    plugin_id: str
    action: str
    state: str
    status: str


# -- Config ---------------------------------------------------------------


class ConfigValueResponse(BaseModel):
    path: str
    type: str
    value: str


class ConfigCurrentResponse(BaseModel):
    content: str
    checksum: str
    path: str
    mtime: str
    entries: list[ConfigValueResponse] = Field(default_factory=list)


class ConfigDocumentResponse(BaseModel):
    content: str
    checksum: str
    path: str
    mtime: str
    data: dict[str, Any] = Field(default_factory=dict)
    redacted_data: dict[str, Any] = Field(default_factory=dict)
    redacted_paths: list[str] = Field(default_factory=list)
    entries: list[ConfigValueResponse] = Field(default_factory=list)


class ConfigSchemaResponse(BaseModel):
    entries: list[dict[str, str]]


class ConfigValidateResponse(BaseModel):
    errors: int
    warnings: int
    ok: bool
    issues: list[dict[str, str]]


class ConfigSaveRequest(BaseModel):
    content: str
    expected_checksum: str
    format: str = "yaml"


class ConfigPatchChange(BaseModel):
    path: str
    value: Any = None
    remove: bool = False
    secret_action: Literal["keep", "replace", "clear"] | None = None


class ConfigPatchRequest(BaseModel):
    expected_checksum: str
    changes: list[ConfigPatchChange] = Field(default_factory=list)


class ConfigSaveResponse(BaseModel):
    saved: bool
    backup_path: str | None = None
    checksum: str = ""
    restart_required: bool = True
    validation: dict[str, Any]


class ConfigRestoreRequest(BaseModel):
    expected_checksum: str | None = None


# -- Sessions -------------------------------------------------------------


class SessionSummaryResponse(BaseModel):
    session_id: str
    session_key_kind: str = ""
    workspace_id: str | None
    created_at: str
    last_active_at: str
    turn_count: int
    metadata: dict = Field(default_factory=dict)


class SessionListResponse(BaseModel):
    sessions: list[SessionSummaryResponse]


class TurnResponse(BaseModel):
    turn_id: int
    role: str
    content: str
    source: str = ""
    created_at: str = ""
    metadata: dict[str, Any] = Field(default_factory=dict)
    sentinel_action: str | None = None
    sentinel_suppressed: bool = False


class SessionHistoryResponse(BaseModel):
    session_id: str
    turns: list[TurnResponse]


class MessageDeliveryResponse(BaseModel):
    delivery_id: str
    target_chat_address: str
    platform: str
    target_type: str
    target_id: str
    source_session_id: str = ""
    source_chat_address: str = ""
    source_user_id: str = ""
    source: str = ""
    delivery_mode: str = ""
    status: str = ""
    message_id: str = ""
    text: str = ""
    error: str = ""
    metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: str = ""
    sentinel_action: str | None = None
    sentinel_suppressed: bool = False


class MessageDeliveryGroupResponse(BaseModel):
    target_chat_address: str
    platform: str
    target_type: str
    target_id: str
    count: int
    last_created_at: str
    last_source: str = ""


class MessageDeliveryGroupsResponse(BaseModel):
    groups: list[MessageDeliveryGroupResponse]


class MessageDeliveriesResponse(BaseModel):
    target_chat_address: str
    deliveries: list[MessageDeliveryResponse]


class SessionSearchResultResponse(BaseModel):
    result_type: Literal["turn", "delivery"]
    id: str
    session_id: str = ""
    target_chat_address: str = ""
    role: str = ""
    source: str = ""
    content: str = ""
    created_at: str = ""
    metadata: dict[str, Any] = Field(default_factory=dict)
    sentinel_action: str | None = None
    sentinel_suppressed: bool = False
    delivery_mode: str = ""
    status: str = ""
    message_id: str = ""


class SessionSearchResponse(BaseModel):
    results: list[SessionSearchResultResponse]


# -- Send Message ---------------------------------------------------------


class SendMessageRequest(BaseModel):
    target: str = ""
    text: str
    session_id: str | None = None


class SendMessageResponse(BaseModel):
    status: str
    session_id: str


# -- Cron -----------------------------------------------------------------


class CronJobResponse(BaseModel):
    job_id: str
    platform: str
    chat_id: str
    mode: str
    prompt: str
    is_active: bool
    next_fire_at: str | None
    run_count: int
    created_at: str
    session_mode: str = "main"
    session_name: str | None = None
    # Extended fields for WebUI
    session_key: str = ""
    chat_type: str = ""
    last_fired_at: str | None = None
    failure_count: int = 0
    last_error: str | None = None
    claimed_at: str | None = None
    workspace_id: str | None = None
    fire_at: str | None = None
    interval_seconds: int | None = None
    cron_expression: str | None = None
    max_runs: int | None = None
    created_by_user_id: str = ""
    created_from_session_id: str = ""
    created_from_chat_address: str = ""
    sender_account_key: str = ""


class CronListResponse(BaseModel):
    jobs: list[CronJobResponse]


class CreateCronRequest(BaseModel):
    target: str = ""
    prompt: str
    mode: Literal["once", "interval", "cron"]
    fire_at: str | None = None
    interval_seconds: int | None = None
    cron_expression: str | None = None
    max_runs: int | None = None
    session_mode: Literal["main", "isolated", "fresh", "named"] = "main"
    session_name: str | None = None

    @model_validator(mode="after")
    def _validate_session(self) -> "CreateCronRequest":
        if self.session_mode == "named":
            if not self.session_name:
                raise ValueError(
                    "session_name is required when session_mode is 'named'"
                )
            if not re.match(r"^[a-zA-Z0-9_-]+$", self.session_name):
                raise ValueError(
                    "session_name must contain only letters, digits, hyphens, and underscores"
                )
        elif self.session_name:
            raise ValueError("session_name is only valid when session_mode is 'named'")
        return self


class CreateCronResponse(BaseModel):
    job_id: str
    status: str


class UpdateCronRequest(BaseModel):
    prompt: str | None = None
    mode: Literal["once", "interval", "cron"] | None = None
    fire_at: str | None = None
    interval_seconds: int | None = None
    cron_expression: str | None = None
    max_runs: int | None = None
    session_mode: Literal["main", "isolated", "fresh", "named"] | None = None
    session_name: str | None = None


class CronActionResponse(BaseModel):
    job_id: str
    status: str


# -- Logs -----------------------------------------------------------------


class LogEntry(BaseModel):
    timestamp: str = ""
    level: str = ""
    logger: str = ""
    event: str = ""
    fields: dict[str, Any] = Field(default_factory=dict)


class LogsResponse(BaseModel):
    entries: list[LogEntry]


# -- Token Usage ----------------------------------------------------------


class TokenTotalsSchema(BaseModel):
    input_tokens: int = 0
    output_tokens: int = 0
    cached_tokens: int = 0
    reasoning_tokens: int = 0
    cache_creation_tokens: int = 0
    estimated_cost: float | None = None
    event_count: int = 0


class ProviderTokenSchema(BaseModel):
    provider_id: str
    model: str
    input_tokens: int = 0
    output_tokens: int = 0
    cached_tokens: int = 0
    reasoning_tokens: int = 0
    cache_creation_tokens: int = 0
    estimated: bool = False
    estimated_cost: float | None = None
    event_count: int = 0


class DailyTokenSchema(BaseModel):
    date: str  # "YYYY-MM-DD"
    input_tokens: int = 0
    output_tokens: int = 0
    provider_id: str = ""
    model: str = ""


class TokenStatsResponse(BaseModel):
    totals: TokenTotalsSchema
    by_provider: list[ProviderTokenSchema]
    daily: list[DailyTokenSchema]


class TokenEventSchema(BaseModel):
    id: int | None = None
    timestamp: str = ""
    session_id: str = ""
    source_tag: str = ""
    provider_id: str = ""
    model: str = ""
    input_tokens: int = 0
    output_tokens: int = 0
    cached_tokens: int = 0
    reasoning_tokens: int = 0
    cache_creation_tokens: int = 0
    estimated: bool = False
    estimated_cost: float | None = None


class TokenEventsResponse(BaseModel):
    events: list[TokenEventSchema]


class TokenClearResponse(BaseModel):
    cleared: bool


# -- Processes ---------------------------------------------------------------


class ProcessInfoResponse(BaseModel):
    name: str
    owner: str
    status: str
    pid: int | None = None
    restart_count: int = 0
    exit_code: int | None = None
    started_at: str | None = None
    last_error: str | None = None
    health: str = "unknown"
    restart_policy: str = "on-failure"
    command: str = ""


class ProcessListResponse(BaseModel):
    processes: list[ProcessInfoResponse]


class ProcessLogsResponse(BaseModel):
    name: str
    stdout: list[str]
    stderr: list[str]


class ProcessActionResponse(BaseModel):
    name: str
    status: str
    ok: bool = True

File: webui/src/api/schemas.ts
export interface ApiError {
  status: number;
  code?: string;
  detail: string;
  requestId?: string;
}

export interface BootstrapResponse {
  app_name: string;
  version: string;
  api_base: string;
  webui_base: string;
  auth: {
    required: boolean;
    mode: "none" | "password" | "bearer";
    api_token_supported?: boolean;
    session_cookie?: boolean;
  };
  features: { id: string; route: string; label: string; scope: string }[];
  server_time: string;
}

export interface AuthSessionResponse {
  authenticated: boolean;
  auth_required: boolean;
  mode: "none" | "password" | "bearer";
  expires_at: string;
}

export interface AuthLoginRequest {
  password: string;
}

export interface AuthLoginResponse {
  authenticated: boolean;
  mode: "password";
  expires_at: string;
}

export interface StatusResponse {
  app: {
    name: string;
    version: string;
    debug: boolean;
    started: boolean;
    started_at: string;
    uptime_seconds: number;
    pid: number;
  };
  resources: {
    cpu_percent: number;
    memory_rss_bytes: number;
    memory_percent: number;
    disk_free_bytes: number;
    db_size_bytes: number;
    workspace_size_bytes: number;
  };
  services: Record<string, string>;
  usage: {
    input_tokens: number;
    output_tokens: number;
    cached_tokens: number;
    reasoning_tokens: number;
    estimated_cost: number | null;
    currency: string | null;
  };
}

export interface ConfigCurrentResponse {
  content: string;
  checksum: string;
  path: string;
  mtime: string;
  entries: { path: string; type: string; value: string }[];
}

export interface ConfigDocumentResponse {
  content: string;
  checksum: string;
  path: string;
  mtime: string;
  data: Record<string, unknown>;
  redacted_data: Record<string, unknown>;
  redacted_paths: string[];
  entries: { path: string; type: string; value: string }[];
}

export interface ConfigSchemaResponse {
  entries: { path: string; type: string; default: string; constraints: string }[];
}

export interface ConfigValidateResponse {
  errors: number;
  warnings: number;
  ok: boolean;
  issues: { severity: string; message: string; path: string }[];
}

export interface CronJob {
  job_id: string;
  platform: string;
  chat_id: string;
  mode: string;
  prompt: string;
  is_active: boolean;
  next_fire_at: string | null;
  run_count: number;
  created_at: string;
  session_mode: string;
  session_name: string | null;
  session_key: string;
  chat_type: string;
  last_fired_at: string | null;
  failure_count: number;
  last_error: string | null;
  claimed_at: string | null;
  workspace_id: string | null;
  fire_at: string | null;
  interval_seconds: number | null;
  cron_expression: string | null;
  max_runs: number | null;
  created_by_user_id: string;
  created_from_session_id: string;
  created_from_chat_address: string;
}

export interface CronListResponse {
  jobs: CronJob[];
}

export interface SessionSummary {
  session_id: string;
  session_key_kind: string;
  workspace_id: string | null;
  created_at: string;
  last_active_at: string;
  turn_count: number;
  metadata: Record<string, unknown>;
}

export interface SessionListResponse {
  sessions: SessionSummary[];
}

export interface Turn {
  turn_id: number;
  role: string;
  content: string;
  source: string;
  created_at: string;
  metadata: Record<string, unknown>;
  sentinel_action: string | null;
  sentinel_suppressed: boolean;
}

export interface SessionHistoryResponse {
  session_id: string;
  turns: Turn[];
}

export interface MessageDelivery {
  delivery_id: string;
  target_chat_address: string;
  platform: string;
  target_type: string;
  target_id: string;
  source_session_id: string;
  source_chat_address: string;
  source_user_id: string;
  source: string;
  delivery_mode: string;
  status: string;
  message_id: string;
  text: string;
  error: string;
  metadata: Record<string, unknown>;
  created_at: string;
  sentinel_action: string | null;
  sentinel_suppressed: boolean;
}

export interface MessageDeliveryGroup {
  target_chat_address: string;
  platform: string;
  target_type: string;
  target_id: string;
  count: number;
  last_created_at: string;
  last_source: string;
}

export interface MessageDeliveryGroupsResponse {
  groups: MessageDeliveryGroup[];
}

export interface MessageDeliveriesResponse {
  target_chat_address: string;
  deliveries: MessageDelivery[];
}

export interface SessionSearchResult {
  result_type: "turn" | "delivery";
  id: string;
  session_id: string;
  target_chat_address: string;
  role: string;
  source: string;
  content: string;
  created_at: string;
  metadata: Record<string, unknown>;
  sentinel_action: string | null;
  sentinel_suppressed: boolean;
  delivery_mode: string;
  status: string;
  message_id: string;
}

export interface SessionSearchResponse {
  results: SessionSearchResult[];
}

export interface FileEntry {
  name: string;
  path: string;
  is_dir: boolean;
  size: number;
  mtime: string;
}

export interface FileListResponse {
  workspace_id: string;
  path: string;
  entries: FileEntry[];
}

export interface FileContentResponse {
  workspace_id: string;
  path: string;
  content: string;
  size: number;
  mtime: string;
}

export interface WorkspaceInfo {
  workspace_id: string;
  is_default: boolean;
  is_active: boolean;
  created_at: string;
  last_active_at: string;
}

export interface WorkspaceListResponse {
  workspaces: WorkspaceInfo[];
  active: string;
}

export interface LogEntry {
  timestamp: string;
  level: string;
  logger: string;
  event: string;
  fields: Record<string, unknown>;
}

export interface LogsResponse {
  entries: LogEntry[];
}

// -- Plugins --

export type PluginState =
  | "found"
  | "loaded"
  | "enabled"
  | "disabled"
  | "error"
  | "unloaded";

export type PluginAction =
  | "load"
  | "enable"
  | "disable"
  | "reload"
  | "unload";

export interface PluginPermissions {
  network?: {
    outbound?: string[];
    inbound?: boolean;
  };
  filesystem?: {
    read?: string[];
    write?: string[];
  };
  memory?: {
    read?: boolean;
    write?: boolean;
  };
  system?: {
    env_vars?: string[];
    subprocess?: boolean;
    signal_handlers?: boolean;
  };
  llm_access?: boolean;
}

export interface PluginCapabilities {
  tools?: Record<string, string>[];
  subscribes_to?: string[];
}

export interface PluginDependency {
  id: string;
  version: string;
}

export interface PluginSummary {
  id: string;
  name: string;
  version: string;
  description: string;
  state: PluginState;
  path: string;
  entrypoint: string;
  load_phase: "pre-agent" | "post-agent" | string;
  nahida_bot_version: string;
  sdk_version: string;
  error_message: string;
  permissions: PluginPermissions;
  capabilities: PluginCapabilities;
  depends_on: PluginDependency[];
  config_keys: string[];
  config_schema: Record<string, unknown>;
  has_config: boolean;
  has_instance: boolean;
  has_runtime_api: boolean;
}

export interface PluginListResponse {
  plugins: PluginSummary[];
}

export interface PluginActionResponse {
  plugin_id: string;
  action: PluginAction;
  state: PluginState;
  status: string;
}

// -- Request / mutation types --

export interface ConfigSaveRequest {
  content: string;
  expected_checksum: string;
  format: string;
}

export interface ConfigPatchChange {
  path: string;
  value?: unknown;
  remove?: boolean;
  secret_action?: "keep" | "replace" | "clear";
}

export interface ConfigPatchRequest {
  expected_checksum: string;
  changes: ConfigPatchChange[];
}

export interface ConfigSaveResponse {
  saved: boolean;
  backup_path: string | null;
  checksum: string;
  restart_required: boolean;
  validation: {
    errors: number;
    warnings: number;
    issues: { severity: string; message: string; path: string }[];
  };
}

export interface SystemActionRequest {
  confirm: boolean;
  reason: string;
}

export interface SystemActionResponse {
  accepted: boolean;
  action: string;
  mode: string;
  message: string;
}

export interface CreateCronRequest {
  target: string;
  prompt: string;
  mode: "once" | "interval" | "cron";
  fire_at?: string | null;
  interval_seconds?: number | null;
  cron_expression?: string | null;
  max_runs?: number | null;
  session_mode: "main" | "isolated" | "fresh" | "named";
  session_name?: string | null;
}

export interface CreateCronResponse {
  job_id: string;
  status: string;
}

export interface UpdateCronRequest {
  prompt?: string | null;
  mode?: "once" | "interval" | "cron" | null;
  fire_at?: string | null;
  interval_seconds?: number | null;
  cron_expression?: string | null;
  max_runs?: number | null;
  session_mode?: "main" | "isolated" | "fresh" | "named" | null;
  session_name?: string | null;
}

export interface CronActionResponse {
  job_id: string;
  status: string;
}

export interface FileWriteRequest {
  path: string;
  content: string;
  workspace_id?: string | null;
}

export interface FileWriteResponse {
  path: string;
  size: number;
  mtime: string;
}

export interface FileUploadRequest {
  path: string;
  file: File;
  workspace_id?: string | null;
  overwrite?: boolean;
}

export interface FileUploadResponse {
  path: string;
  size: number;
  mtime: string;
}

export interface FileCreateRequest {
  path: string;
  content: string;
  workspace_id?: string | null;
}

export interface FileRenameRequest {
  path: string;
  new_name: string;
  workspace_id?: string | null;
}

export interface FileRenameResponse {
  path: string;
}

export interface FileDeleteRequest {
  path: string;
  workspace_id?: string | null;
}

export interface FileDeleteResponse {
  status: string;
  trash_path: string;
}

// -- Token Usage --

export interface TokenTotals {
  input_tokens: number;
  output_tokens: number;
  cached_tokens: number;
  reasoning_tokens: number;
  cache_creation_tokens: number;
  estimated_cost: number | null;
  event_count: number;
}

export interface ProviderToken {
  provider_id: string;
  model: string;
  input_tokens: number;
  output_tokens: number;
  cached_tokens: number;
  reasoning_tokens: number;
  cache_creation_tokens: number;
  estimated: boolean;
  estimated_cost: number | null;
  event_count: number;
}

export interface DailyToken {
  date: string;
  input_tokens: number;
  output_tokens: number;
  provider_id: string;
  model: string;
}

export interface TokenStatsResponse {
  totals: TokenTotals;
  by_provider: ProviderToken[];
  daily: DailyToken[];
}

export interface TokenEvent {
  id: number | null;
  timestamp: string;
  session_id: string;
  source_tag: string;
  provider_id: string;
  model: string;
  input_tokens: number;
  output_tokens: number;
  cached_tokens: number;
  reasoning_tokens: number;
  cache_creation_tokens: number;
  estimated: boolean;
  estimated_cost: number | null;
}

export interface TokenEventsResponse {
  events: TokenEvent[];
}

export interface TokenClearResponse {
  cleared: boolean;
}
